File: 46.全排列/permute.go
package problem0046

// 回溯算法的框架
// func backtrack(路径, 选择列表) {
//   if 满足结束条件 {
//     result = append(result, 路径)
//     return
//   }
// 	 for 选择 in 选择列表 {
// 	   做选择
//     backtrack(路径，选择列表)
//     撤销选择
// 	 }
// }
func permute(nums []int) [][]int {
	var (
		result    [][]int
		used      = make([]bool, len(nums))
		backTrack func(path []int)
	)
	backTrack = func(path []int) {
		if len(path) == len(nums) {
			temp := make([]int, len(path))
			copy(temp, path)
			result = append(result, temp)
			return
		}
		for i := 0; i < len(nums); i++ {
			if used[i] {
				continue
			}
			path = append(path, nums[i])
			used[i] = true
			backTrack(path)
			used[i] = false
			path = path[:len(path)-1]
		}
	}
	backTrack([]int{})
	return result
}

File: 46.全排列/permute.js
// 空间复杂度 O(n)
// 时间复杂度 O(n*n!)：backTrack的调用次数是O(n!)的，每次需要将当前答案使用 O(n) 的时间复制到答案数组中
function permute(nums) {
    const result = []
    const N = nums.length
    const visit = new Array(N).fill(false)

    const backTrack = (combine) => {
        if (combine.length === N) {
            result.push(combine)
            return
        }
        for (let i = 0; i < N; i++) {
            if (visit[i]) {
                continue
            }
            visit[i] = true
            combine.push(nums[i])
            backTrack([...combine])
            combine.pop()
            visit[i] = false
        }
    }

    backTrack([])

    return result
}